core/src/hir/paths.rs

```diff
@@ -0,0 +1,59 @@
+use super::{EnumId, OpaqueId, OutStructId, Ref, StructId, TypeLifetimes};
+
+/// Path to a struct that may appear as an output.
+pub enum ReturnableStruct {
+    Struct(Struct),
+    OutStruct(OutStruct),
+}
+
+/// Path to a struct that can only be used as an output.
+pub struct OutStruct {
+    pub lifetimes: TypeLifetimes,
+    pub(crate) tcx_id: OutStructId,
+}
+
+/// Path to a struct that can be used in inputs and outputs.
+pub struct Struct {
+    pub lifetimes: TypeLifetimes,
+    pub(crate) tcx_id: StructId,
+}
+
+/// Path to an opaque that may be owned (`Box<T>`) or borrowed (`&T`), and may
+/// also be nullable (`Option<Box<T>>`/`Option<&T>`).
+pub struct OutOpaque {
+    pub lifetimes: TypeLifetimes,
+    pub nullable: bool,
+    pub borrow: Ownership,
+    pub(crate) tcx_id: OpaqueId,
+}
+
+/// Path to an opaque that is borrowed (`&T`), and may
+/// also be nullable (`Option<&T>`).
+pub struct Opaque {
+    pub lifetimes: TypeLifetimes,
+    pub nullable: bool,
+    pub borrow: Ref,
+    pub(crate) tcx_id: OpaqueId,
+}
+
+/// Path to an opaque that can appear in the `&self` position and is always borrowed
+/// (`&T`), but not nullable.
+pub struct SelfOpaque {
```

Comment:
thought: this might be _too much_ of a distinction, but we can see and merge it later if we need

Reply:
What if we make a `OpaquePath` type that has two generic arguments. So the `SelfType` could have an `Opaque(OpaquePath<NonNull, Borrow>)` variant, `Type` could have `Opaque(OpaquePath<Optional, Borrow>)`, and `OutType` could have `Opaque(OpaquePath<Optional, MaybeOwned>)`? Then we could implement different functions depending on which types its generic over.

This would allow us to show that they're all basically the same thing with slightly different invariants, and the compiler with monomorphize them into what we currently have anyways.

Reply:
That might work! Let's try that in a followup though.